nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-services/src/main/java/org/apache/nifi/mongodb/MongoDBLookupService.java

```diff
@@ -52,68 +54,125 @@
-            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
-            .required(false)
-            .build();
+        .name("mongo-lookup-value-field")
+        .displayName("Lookup Value Field")
+        .description("The field whose value will be returned when the lookup key(s) match a record. If not specified then the entire " +
+                "MongoDB result document minus the _id field will be returned as a record.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .build();
+    public static final PropertyDescriptor PROJECTION = new PropertyDescriptor.Builder()
+        .name("mongo-lookup-projection")
+        .displayName("Projection")
+        .description("Specifies a projection for limiting which fields will be returned.")
+        .required(false)
+        .build();
 
     private String lookupValueField;
 
-    private static final List<PropertyDescriptor> lookupDescriptors;
-
-    static {
-        lookupDescriptors = new ArrayList<>();
-        lookupDescriptors.addAll(descriptors);
-        lookupDescriptors.add(LOOKUP_VALUE_FIELD);
-    }
-
     @Override
     public Optional<Object> lookup(Map<String, Object> coordinates) throws LookupFailureException {
-        Map<String, Object> clean = new HashMap<>();
-        clean.putAll(coordinates);
+        Map<String, Object> clean = coordinates.entrySet().stream()
+            .filter(e -> !schemaNameProperty.equals(String.format("${%s}", e.getKey())))
+            .collect(Collectors.toMap(
+                e -> e.getKey(),
+                e -> e.getValue()
+            ));
         Document query = new Document(clean);
 
         if (coordinates.size() == 0) {
             throw new LookupFailureException("No keys were configured. Mongo query would return random documents.");
         }
 
         try {
-            Document result = this.findOne(query);
+            Document result = projection != null ? controllerService.findOne(query, projection) : controllerService.findOne(query);
 
             if(result == null) {
                 return Optional.empty();
             } else if (!StringUtils.isEmpty(lookupValueField)) {
                 return Optional.ofNullable(result.get(lookupValueField));
             } else {
-                final List<RecordField> fields = new ArrayList<>();
+                RecordSchema schema = loadSchema(coordinates);
 
-                for (String key : result.keySet()) {
-                    if (key.equals("_id")) {
-                        continue;
-                    }
-                    fields.add(new RecordField(key, RecordFieldType.STRING.getDataType()));
-                }
-
-                final RecordSchema schema = new SimpleRecordSchema(fields);
-                return Optional.ofNullable(new MapRecord(schema, result));
+                RecordSchema toUse = schema != null ? schema : convertSchema(result);
```

Comment:
@mattyb149 I think the answer to your last question is here. If you specify `schema.name` in the coordinates, it'll get that from `loadSchema`. If not, it calls `convertSchema`. The rest of the lookup strategies don't make much sense in this case so I can back out the change to extended `SchemaRegistryService` if that makes sense.

Reply:
One thing you could do is to override getSupportedPropertyDescriptors() and add your own property for Schema Access Strategy that only has the relevant ones, including your own strategy of "Infer Schema From Document" or something.

Reply:
That's probably the right way to do it because we should have it blow up if it can't get the schema on the first pass instead of silently falling back onto the inference option. Once I get that worked out, I'll copy pasta it over the ES one as well.